pkg/agent/controller/networkpolicy/networkpolicy_controller.go

```diff
@@ -538,6 +540,33 @@ func NewNetworkPolicyController(antreaClientGetter client.AntreaClientProvider,
 	return c, nil
 }
 
+func (c *Controller) GetFqdnCache(fqdnFilter querier.FQDNCacheFilter) []types.DnsCacheEntry {
+	cacheEntryList := []types.DnsCacheEntry{}
+	var pattern *regexp.Regexp
+	var err error
+	if fqdnFilter != (querier.FQDNCacheFilter{}) {
+		// have to convert human readable regex, i.e. *.example.com into regex that can be used
+		regexPattern := "^" + strings.ReplaceAll(regexp.QuoteMeta(fqdnFilter.DomainName), `\*`, ".*") + "$"
+		pattern, err = regexp.Compile(regexPattern)
+		if err != nil {
+			// this pattern will match no strings if there is an error with the regex formatting or usage with the user specified --domain flag
+			pattern = regexp.MustCompile(`a\A`)
+		}
+	} else {
+		// this pattern will match all strings if the filter is unset
+		pattern = regexp.MustCompile(`.*`)
+	}
+	for fqdn, dnsMeta := range c.fqdnController.dnsEntryCache {
+		for _, ipWithExpiration := range dnsMeta.responseIPs {
+			if fqdnFilter == (querier.FQDNCacheFilter{}) || pattern.MatchString(fqdn) {
```

Comment:
fqdnFilter == nil will also be a much more appropriate check here to avoid constructing an empty `FQDNCacheFilter` for each dns data in cache.